core/src/main/java/org/elasticsearch/index/query/PercolatorQuery.java

```diff
@@ -164,52 +163,42 @@ public Scorer scorer(LeafReaderContext leafReaderContext) throws IOException {
                 }
 
                 final QueryRegistry.Leaf percolatorQueries = queryRegistry.getQueries(leafReaderContext);
-                return new Scorer(this) {
-
-                    @Override
-                    public DocIdSetIterator iterator() {
-                        return TwoPhaseIterator.asDocIdSetIterator(twoPhaseIterator());
-                    }
-
-                    @Override
-                    public TwoPhaseIterator twoPhaseIterator() {
-                        return new TwoPhaseIterator(approximation.iterator()) {
-                            @Override
-                            public boolean matches() throws IOException {
-                                return matchDocId(approximation.docID());
+                if (needsScores) {
```

Comment:
> We could just kep things like before and insert if (needsScores) where necessary?

I think that now that the percolator has been improved the amount of docs going to be evaluated is going to increase and it feels to me that this specializing is justified and also the the extra code for this specializing is contained inside this class.

> Also we could just call TopDocs topDocs = searcher.search(query, 1) instead of the specialized collector?

The reason I created this collector was because `searcher.search(query, 1)` is doing too much in the case for the MemoryIndex. (creating a search manager, priority queues etc. which makes sense in all other cases). I'm ok with removing this. I think instead we should improve the `MemoryIndex#createSearcher()` the return an specialized IndexSearcher impl that does what make sense in the context of the a single document index.